File: 022_GenerateParentheses.py
# Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

 

# Example 1:

# Input: n = 3
# Output: ["((()))","(()())","(())()","()(())","()()()"]
# Example 2:

# Input: n = 1
# Output: ["()"]


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def backtrack(string, left, right):
            if len(string) == 2 * n:
                result.append(string)
                return
            if(left < n):
                backtrack(string+'(', left + 1, right)
            if(right < left):
                backtrack(string+')', left, right+1)
        result = []
        backtrack('', 0, 0 )
        return result

File: 022_GenerateParentheses.cpp
// Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

 

// Example 1:

// Input: n = 3
// Output: ["((()))","(()())","(())()","()(())","()()()"]
// Example 2:

// Input: n = 1
// Output: ["()"]

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<std::string> result;
        backtrack("",0,0,n,result);
        return result;
    }

private:
    void backtrack(std::string s, int left, int right, int n, vector<std::string>& result){
        if (s.length() == 2 * n){
            result.push_back(s);
            return;
        }

        if(left < n) backtrack(s+'(', left + 1, right, n, result);
        if(right<left) backtrack(s+')', left, right+1, n , result);
    }
};
